sda/router.go

```diff
@@ -0,0 +1,543 @@
+	var err error
+	var c network.SecureConn
+	// try to open connection
+	c, err = t.host.Open(id)
+	if err != nil {
+		return nil, err
+	}
+	log.Lvl3("Host", t.workingAddress, "connected to", c.Remote())
+	t.registerConnection(c)
+	go t.handleConn(c)
+	return c, nil
+}
+
+// Close shuts down all network connections and closes the listener.
+func (t *TCPRouter) Close() error {
+	t.closingMut.Lock()
+	defer t.closingMut.Unlock()
+	if t.isClosing {
+		return errors.New("Already closing")
+	}
+	log.Lvl4(t.serverIdentity.First(), "Starts closing")
+	t.isClosing = true
+	// stop the Run
+	t.closing <- true
+	t.networkLock.Lock()
+	if t.processMessagesStarted {
+		// Tell ProcessMessages to quit
+		close(t.quitProcessMsg)
+	}
+	t.networkLock.Unlock()
+	// The tcp host is supposed to take care of the connection for us
+	if err := t.host.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// closeConnection closes a connection and removes it from the connections-map
+// The t.networkLock must be taken.
+func (t *TCPRouter) closeConnection(c network.SecureConn) error {
+	t.networkLock.Lock()
+	defer t.networkLock.Unlock()
+	log.Lvl4(t.serverIdentity.First(), "Closing connection", c, c.Remote(), c.Local())
+	err := c.Close()
+	if err != nil {
+		return err
+	}
+	delete(t.connections, c.ServerIdentity().ID)
+	return nil
+}
+
+// Handle a connection => giving messages to the MsgChans
+func (t *TCPRouter) handleConn(c network.SecureConn) {
+	address := c.Remote()
+	for {
+		ctx := context.TODO()
+		am, err := c.Receive(ctx)
+		// So the receiver can know about the error
+		am.SetError(err)
+		am.From = address
+		log.Lvl5(t.workingAddress, "Got message", am)
+		if err != nil {
+			t.closingMut.Lock()
+			log.Lvlf4("%+v got error (%+s) while receiving message (isClosing=%+v)",
+				t.serverIdentity.First(), err, t.isClosing)
+			t.closingMut.Unlock()
+			if err == network.ErrClosed || err == network.ErrEOF || err == network.ErrTemp {
+				log.Lvl4(t.serverIdentity.First(), c.Remote(), "quitting handleConn for-loop", err)
+				t.closeConnection(c)
+				return
+			}
+			log.Error(t.serverIdentity.Addresses, "Error with connection", address, "=>", err)
+		} else {
+			t.closingMut.Lock()
+			if !t.isClosing {
+				log.Lvl5(t.workingAddress, "Send message to networkChan", len(t.networkChan))
+				t.networkChan <- am
+			}
+			t.closingMut.Unlock()
+		}
+	}
+}
+
+// registerConnection registers an ServerIdentity for a new connection, mapped with the
+// real physical address of the connection and the connection itself
+// it locks (and unlocks when done):  networkLock
+func (t *TCPRouter) registerConnection(c network.SecureConn) {
+	log.Lvl4(t.serverIdentity.First(), "registers", c.ServerIdentity().First())
+	t.networkLock.Lock()
+	defer t.networkLock.Unlock()
+	id := c.ServerIdentity()
+	_, okc := t.connections[id.ID]
+	if okc {
+		// TODO - we should catch this in some way
```

Comment:
I'm not sure. I've put the TODO here a long time ago and it has never re appeared again so... I'm not sure it's useful. Moreover, we should before doing anything else about this, specify what happens when the same connection or identity appears again. For the moment the latest registered connection will always be the one active. I guess that's subject to discussion somewhere else... (and things will change when/if we introduce TLS in the whole story)